index.js

```diff
@@ -1,201 +1,114 @@
+    /* H A N D L E R S */
 
-        invariant(
-            type === NOTIFICATION_RECEIVED_EVENT ||
-            type === NOTIFICATION_OPENED_EVENT ||
-            type === IDS_AVAILABLE_EVENT ||
-//            type === SUBSCRIPTION_EVENT ||
-//            type === PERMISSION_EVENT ||
-            type === EMAIL_SUBSCRIPTION_EVENT ||
-            type === IN_APP_MESSAGE_CLICKED_EVENT,
-            'OneSignal only supports received, opened, ids, emailSubscription, and inAppMessageClicked events'
-        );
-
-        _eventTypeHandler.delete(type);
+    /**
+     * Sets the handler that fires before the notification is displayed
+     * Callback parameter is a `NotificationReceivedEvent` with:
+     *  - notification data
+     *  - `complete` function that accepts the `NotificationReceivedEvent`
+     * @param  {function} handler
+     */
+    static setNotificationWillShowInForegroundHandler(handler){
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(handler);
+        RNOneSignal.setNotificationWillShowInForegroundHandler();
+        eventManager.setEventHandler(NOTIFICATION_WILL_SHOW, handler);
     }
 
-    static clearListeners() {
-        if (!checkIfInitialized()) return;
+    static setNotificationOpenedHandler(handler){
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(handler);
 
-        for(var i = 0; i < _eventNames.length; i++) {
-            _listeners[_eventNames].remove();
-        }
+        RNOneSignal.setNotificationOpenedHandler();
+        eventManager.setEventHandler(NOTIFICATION_OPENED, handler);
     }
 
+    /* R E G I S T R A T I O N  E T C */
+
     static registerForPushNotifications() {
-        if (!checkIfInitialized()) return;
+        if (!checkIfInitialized(RNOneSignal)) return;
 
         if (Platform.OS === 'ios') {
             RNOneSignal.registerForPushNotifications();
         } else {
-            console.log("This function is not supported on Android");
+            console.log("registerForPushNotifications: this function is not supported on Android");
         }
     }
 
-    static promptForPushNotificationsWithUserResponse(callback) {
-        if (!checkIfInitialized()) return;
+    static promptForPushNotificationsWithUserResponse(handler) {
+        if (!checkIfInitialized(RNOneSignal)) return;
 
         if (Platform.OS === 'ios') {
-            invariant(
-                typeof callback === 'function',
-                'Must provide a valid callback'
-            );
-            RNOneSignal.promptForPushNotificationsWithUserResponse(callback);
+            isValidCallback(handler);
+            RNOneSignal.promptForPushNotificationsWithUserResponse(handler);
         } else {
-            console.log("This function is not supported on Android");
+            console.log("promptForPushNotificationsWithUserResponse: this function is not supported on Android");
         }
     }
 
     static requestPermissions(permissions) {
```

Comment:
Remove `requestPermissions` method, it does the same thing as `promptForPushNotificationsWithUserResponse`

Reply:
Done